Add LLaMA perf benchmark workflow for Apple iOS

huydhn opened the pull request:
This PR setups the framework to run LLaMA perf benchmark on Apple iOS:

* LLaMAPerfBenchmarkApp is an empty app just to host the benchmark on device.  We could extend this later to make it the generic perf benchmark loader app for iOS
* The model (llama2) is benchmarked in `testLlama2`.  I follow the same example from ET demo app where another model (mv3) is bundled together with the test suite

One caveat that I need to figure out is how to get the benchmark results out. Atm, I use this script https://github.com/ChrisMash/XCResultExtractor to extract it manually from `xcresult` file. For example,

```
$ python xcresult_extractor.py --file /Users/huydo/Library/Developer/Xcode/DerivedData/LLaMA-bchtdulhbuqeyparjpfpuchdxbva/Logs/Test/Test-LLaMAPerfBenchmark-2024.08.29_14-37-08--0700.xcresult

Test Suite 'LLaMAPerfBenchmarkTests' started at 2024-08-29 14:38:54.228.
Test Case '-[LLaMAPerfBenchmarkTests.LLaMAPerfBenchmarkTests testLlama2]' started.
2024-08-29 14:38:54.230612-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:00.559179042000000 executorch:runner.cpp:69] Creating LLaMa runner: model_path=/Users/huydo/Library/Developer/XCTestDevices/141725A5-9829-4738-95E4-5CF3A5688020/data/Containers/Bundle/Application/0EC501AC-B18F-4303-A98D-35D3DEBDDAE6/LLaMAPerfBenchmark.app/PlugIns/LLaMAPerfBenchmarkTests.xctest/llam$
2024-08-29 14:38:54.230715-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:00.559179 executorch:runner.cpp:69] Creating LLaMa runner: model_path=/Users/huydo/Library/Developer/XCTestDevices/141725A5-9829-4738-95E4-5CF3A5688020/data/Containers/Bundle/Application/0EC501AC-B18F-4303-A98D-35D3DEBDDAE6/LLaMAPerfBenchmark.app/PlugIns/LLaMAPerfBenchmarkTests.xctest/llam$
2024-08-29 14:38:54.721031-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:01.1049707332000000 executorch:runner.cpp:85] Reading metadata from model
2024-08-29 14:38:54.721134-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:01.049707 executorch:runner.cpp:85] Reading metadata from model
2024-08-29 14:38:54.721690-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:01.1050379124000000 executorch:runner.cpp:110] Metadata: get_vocab_size = 32000
2024-08-29 14:38:54.721739-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:01.050379 executorch:runner.cpp:110] Metadata: get_vocab_size = 32000
2024-08-29 14:38:54.721794-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:01.1050485249000000 executorch:runner.cpp:110] Metadata: get_bos_id = 1
2024-08-29 14:38:54.721828-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:01.050485 executorch:runner.cpp:110] Metadata: get_bos_id = 1
2024-08-29 14:38:54.721874-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:01.1050565666000000 executorch:runner.cpp:110] Metadata: use_sdpa_with_kv_cache = 1
2024-08-29 14:38:54.721923-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:01.050565 executorch:runner.cpp:110] Metadata: use_sdpa_with_kv_cache = 1
2024-08-29 14:38:54.721993-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:01.1050680207000000 executorch:runner.cpp:110] Metadata: get_n_eos = 1
2024-08-29 14:38:54.722034-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:01.050680 executorch:runner.cpp:110] Metadata: get_n_eos = 1
2024-08-29 14:38:54.722079-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:01.1050770707000000 executorch:runner.cpp:110] Metadata: append_eos_to_prompt = 0
2024-08-29 14:38:54.722120-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:01.050770 executorch:runner.cpp:110] Metadata: append_eos_to_prompt = 0
2024-08-29 14:38:54.722167-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:01.1050857957000000 executorch:runner.cpp:110] Metadata: get_max_seq_len = 128
2024-08-29 14:38:54.722578-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:01.050857 executorch:runner.cpp:110] Metadata: get_max_seq_len = 128
2024-08-29 14:38:54.722633-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:01.1051322416000000 executorch:runner.cpp:110] Metadata: enable_dynamic_shape = 1
2024-08-29 14:38:54.722677-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:01.051322 executorch:runner.cpp:110] Metadata: enable_dynamic_shape = 1
2024-08-29 14:38:54.722719-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:01.1051410124000000 executorch:runner.cpp:110] Metadata: use_kv_cache = 1
2024-08-29 14:38:54.722755-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:01.051410 executorch:runner.cpp:110] Metadata: use_kv_cache = 1
2024-08-29 14:38:54.722805-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:01.1051494791000000 executorch:runner.cpp:110] Metadata: get_n_bos = 1
2024-08-29 14:38:54.722842-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:01.051494 executorch:runner.cpp:110] Metadata: get_n_bos = 1
2024-08-29 14:38:54.722887-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:01.1051578374000000 executorch:runner.cpp:117] eos_id = 2
2024-08-29 14:38:54.722924-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:01.051578 executorch:runner.cpp:117] eos_id = 2
2024-08-29 14:38:55.363728-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:01.1692404707000000 executorch:text_prefiller.cpp:57] Prefill token result numel(): 32000
2024-08-29 14:38:55.363829-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:01.692404 executorch:text_prefiller.cpp:57] Prefill token result numel(): 32000
How do you do! I'm testing llama2 on mobile device."
Gretlasses, a yummy-looking tool, still beaming.
"Can I have one?" asked the little girl.
Gretilla nodded, and handed her sister a spinning glass.
Gretilla grabbed the glass and started to test it.
"Ah, it's so pretty!" exclaimed the little girl.
Gretilla smiled and said, "Yes, it's very pretty. Now, let's take a break and enjoy this yummy glass together!"
They both
PyTorchObserver {"prompt_tokens":16,"generated_tokens":111,"model_load_start_ms":0,"model_load_end_ms":0,"inference_start_ms":1724967534723,"inference_end_ms":1724967546771,"prompt_eval_end_ms":1724967535364,"first_token_ms":1724967535364,"aggregate_sampling_time_ms":62,"SCALING_FACTOR_UNITS_PER_SECOND":1000}
2024-08-29 14:39:06.771312-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:13.13099907195000000 executorch:stats.h:84] 	Prompt Tokens: 16    Generated Tokens: 111
2024-08-29 14:39:06.771428-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:13.099907 executorch:stats.h:84] 	Prompt Tokens: 16    Generated Tokens: 111
2024-08-29 14:39:06.771492-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:13.13100095862000000 executorch:stats.h:90] 	Model Load Time:		0.000000 (seconds)
2024-08-29 14:39:06.771542-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:13.100095 executorch:stats.h:90] 	Model Load Time:		0.000000 (seconds)
2024-08-29 14:39:06.771583-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:13.13100188904000000 executorch:stats.h:100] 	Total inference time:		12.048000 (seconds)		 Rate: 	9.213147 (tokens/second)
2024-08-29 14:39:06.771628-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:13.100188 executorch:stats.h:100] 	Total inference time:		12.048000 (seconds)		 Rate: 	9.213147 (tokens/second)
2024-08-29 14:39:06.771678-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:13.13100283195000000 executorch:stats.h:108] 		Prompt evaluation:	0.641000 (seconds)		 Rate: 	24.960998 (tokens/second)
2024-08-29 14:39:06.771733-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:13.100283 executorch:stats.h:108] 		Prompt evaluation:	0.641000 (seconds)		 Rate: 	24.960998 (tokens/second)
2024-08-29 14:39:06.771785-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:13.13100392487000000 executorch:stats.h:119] 		Generated 111 tokens:	11.407000 (seconds)		 Rate: 	9.730867 (tokens/second)
2024-08-29 14:39:06.771855-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:13.100392 executorch:stats.h:119] 		Generated 111 tokens:	11.407000 (seconds)		 Rate: 	9.730867 (tokens/second)
2024-08-29 14:39:06.771890-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:13.13100496737000000 executorch:stats.h:127] 	Time to first generated token:	0.641000 (seconds)
2024-08-29 14:39:06.771980-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:13.100496 executorch:stats.h:127] 	Time to first generated token:	0.641000 (seconds)
2024-08-29 14:39:06.772028-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:13.13100634320000000 executorch:stats.h:134] 	Sampling time over 127 tokens:	0.062000 (seconds)
2024-08-29 14:39:06.772114-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:13.100634 executorch:stats.h:134] 	Sampling time over 127 tokens:	0.062000 (seconds)
Test Case '-[LLaMAPerfBenchmarkTests.LLaMAPerfBenchmarkTests testLlama2]' passed (12.654 seconds).
Test Suite 'LLaMAPerfBenchmarkTests' passed at 2024-08-29 14:39:06.883.
	 Executed 1 test, with 0 failures (0 unexpected) in 12.654 (12.655) seconds
```

### Testing

https://github.com/pytorch/executorch/actions/runs/10639727158/job/29498590702

huydhn committed:
Add LLaMA perf benchmark workflow for Apple iOS

huydhn committed:
Export models

huydhn committed:
Run with CONDA_RUN

huydhn committed:
Fix wrong secret

huydhn committed:
Still wrong secret

huydhn committed:
Add a benchmark test for iOS LLaMA app

huydhn committed:
Run on AWS

huydhn committed:
Build for testing

huydhn committed:
Fix model path

huydhn committed:
Lower target to 17.0

huydhn committed:
Try a different way to set macros

huydhn committed:
Create a new LLaMAPerfBenchmark app

huydhn committed:
Fix lint

huydhn committed:
Clean up redundant changes

huydhn committed:
Fix wrong app name

huydhn committed:
Remove debug statement

guangy10 commented:
Thank you for putting this up. I will review it on Fri. 

Just a few quick update about how we would build the generic app for iOS and how it would be integrated: 

> LLaMAPerfBenchmarkApp is an empty app just to host the benchmark on device. We could extend this later to make it the generic perf benchmark loader app for iOS

@shoumikhin You may want to take a look at this PR about how the empty app is integrated since you are working on the generic app for iOS.

The model (llama2) is benchmarked in testLlama2. I follow the same example from ET demo app where another model (mv3) is bundled together with the test suite

@huydhn Regarding the bundled model to the app, I think @shoumikhin is planning to create a generic app for iOS w/o bundling any model to the app. Instead it will locate the model from local fs. @shoumikhin can confirm. It will be same as how the app works in the Android workflow. It means, we will need to extend the test-spec to make deploy the model to some location where the app can loop up from.

> One caveat that I need to figure out is how to get the benchmark results out. Atm, I use this script https://github.com/ChrisMash/XCResultExtractor to extract it manually from xcresult file.

@shoumikhin can definitely provide feedback for this. He also has a plan to always build framework on the fly to ensure the latest kernels/backends changes can be reflected in the app. (Currently we are using pre-build framework)

huydhn commented:
@guangy10 @shoumikhin Yeah, please let me know if you have other approaches in mind to copy the model to device and get the results back.  Here are some extra steps that I'm looking at:

* For iOS, the extra data is uploaded to the application sandbox on device. Let me try to read the files from that instead, then we will be able to do the same for iOS what we have for Android.
* The iOS Llama runner logs some benchmark information, but doesn't expose it anywhere else.  So, I guess the runner code needs to be updated to return these information.  This could be done in a later PR.

```
{"prompt_tokens":16,"generated_tokens":111,"model_load_start_ms":0,"model_load_end_ms":0,"inference_start_ms":1724967534723,"inference_end_ms":1724967546771,"prompt_eval_end_ms":1724967535364,"first_token_ms":1724967535364,"aggregate_sampling_time_ms":62,"SCALING_FACTOR_UNITS_PER_SECOND":1000}
```

> @shoumikhin can definitely provide feedback for this. He also has a plan to always build framework on the fly to ensure the latest kernels/backends changes can be reflected in the app. (Currently we are using pre-build framework)

Getting pre-build framework here in CI is not the right setup indeed.  I notice the same when @shoumikhin told me to clean up the local cache to fix the mismatch checksum of pre-build framework binaries :)

huydhn committed:
Attempt to read model from app sandbox

huydhn committed:
Add debug pull_request

huydhn committed:
Try the app main bundle

guangy10 commented on .github/workflows/apple-perf.yml:
```diff
@@ -0,0 +1,313 @@
+        GITHUB_RUNNER=1 PYTHON_EXECUTABLE=python ${CONDA_RUN} --no-capture-output \
+          .ci/scripts/setup-macos.sh "${BUILD_TOOL}"
+
+        if [[ ${{ matrix.delegate }} == "qnn" ]]; then
+          PYTHON_EXECUTABLE=python ${CONDA_RUN} --no-capture-output bash \
+            .ci/scripts/setup-qnn-deps.sh
+          PYTHON_EXECUTABLE=python ${CONDA_RUN} --no-capture-output bash \
+            .ci/scripts/build-qnn-sdk.sh
+        fi
```
QNN is android only. Can be removed here.

guangy10 commented on .github/workflows/apple-perf.yml:
```diff
@@ -0,0 +1,313 @@
+            bash .ci/scripts/test_llama.sh "${{ matrix.model }}" "${BUILD_MODE}" "${DTYPE}" "${DELEGATE_CONFIG}" "${ARTIFACTS_DIR_NAME}"
+        else
+          PYTHON_EXECUTABLE=python ${CONDA_RUN} --no-capture-output \
+            bash .ci/scripts/test.sh "${{ matrix.model }}" "${BUILD_MODE}" "${{ matrix.delegate }}" "${ARTIFACTS_DIR_NAME}"
```
Neither coreml nor mps are yet supported in this script

huydhn replied:
I see, let me put a `TODO` comment here for now

guangy10 commented on .github/workflows/apple-perf.yml:
```diff
@@ -0,0 +1,313 @@
+      submodules: 'true'
+      ref: ${{ github.event_name == 'pull_request' && github.event.pull_request.head.sha || github.sha }}
+      upload-artifact: ios-apps
+      secrets-env: BUILD_CERTIFICATE_BASE64 BUILD_PROVISION_PROFILE_BASE64 KEYCHAIN_PASSWORD
```
👍

guangy10 commented on .github/workflows/apple-perf.yml:
```diff
@@ -0,0 +1,313 @@
+          backends/apple/mps/install_requirements.sh
+        echo "::endgroup::"
+
+        ${CONDA_RUN} --no-capture-output \
+          build/build_apple_llm_demo.sh ${{ matrix.tokenizer }} ${ARTIFACTS_DIR_NAME}
```
@shoumikhin FYI, once we have the generic iOS app, this is where it will be integrated to with its own build script

guangy10 commented on .github/workflows/apple-perf.yml:
```diff
@@ -0,0 +1,313 @@
+      project-arn: arn:aws:devicefarm:us-west-2:308535385114:project:02a2cf0f-6d9b-45ee-ba1a-a086587469e6
+      device-pool-arn: ${{ matrix.device }}
+      # Uploaded to S3 from the previous job
+      ios-ipa-archive: https://gha-artifacts.s3.amazonaws.com/${{ github.repository }}/${{ github.run_id }}/artifact/LLaMAPerfBenchmark.ipa
+      ios-xctestrun-zip: https://gha-artifacts.s3.amazonaws.com/${{ github.repository }}/${{ github.run_id }}/artifact/LLaMAPerfBenchmark.xctestrun.zip
+      test-spec: ${{ inputs.test_spec || 'https://ossci-ios.s3.amazonaws.com/executorch/default-ios-device-farm-appium-test-spec.yml' }}
+      extra-data: https://gha-artifacts.s3.amazonaws.com/${{ github.repository }}/${{ github.run_id }}/artifact/${{ matrix.model }}_${{ matrix.delegate }}/model.zip
```
This setup assumes the model is not bundled in the app, so the test-spect must guide the app where to locate the models.

huydhn replied:
I'm testing this out atm.  AWS doc is lacking here.  My understanding is that the extra data will be there in the app container, I need to figure out where.  I'm currently seeing them under:

```
LLaMAPerfBenchmark.app/aatp/data/
LLaMAPerfBenchmark.app/aatp/data/llama2.pte  
LLaMAPerfBenchmark.app/aatp/data/tokenizer.bin
```

huydhn replied:
It's finally working https://github.com/pytorch/executorch/actions/runs/10639727158/job/29498590702?pr=4953, so the extra data archive is copied to the above location, and can be accessed using something like:

```
let modelPath = Bundle.main.path(
  forResource: "llama2",
  ofType: "pte",
  inDirectory: "aatp/data" <-- This path is important and is not documented anywhere lol
)
```

The good news is the current spec works, no need to change anything there.

guangy10 replied:
cc: @shoumikhin

guangy10 commented on build/build_apple_llm_demo.sh:
```diff
@@ -0,0 +1,73 @@
+TOKENIZER="${1:-bpe}"
+ARTIFACTS_DIR_NAME="$2"
+
+APP_PATH="examples/demo-apps/apple_ios/LLaMA/LLaMA"
```
I will suggest host the new generic benchmarking app under `extension/apple/benchmark` and give it a better name. cc: @shoumikhin

guangy10 commented on build/build_apple_llm_demo.sh:
```diff
@@ -0,0 +1,73 @@
+
+APP_PATH="examples/demo-apps/apple_ios/LLaMA/LLaMA"
+
+if [[ "${TOKENIZER}" = "bpe" ]]; then
```
This is very llm specific

huydhn replied:
Yeah, I need this mainly to overwrite the default `ET_USE_TIKTOKEN=1` set throughout the app.  This might go away once we have a better generic iOS app at hand.

guangy10 replied:
cc: @shoumikhin

guangy10 commented on build/build_apple_llm_demo.sh:
```diff
@@ -0,0 +1,73 @@
+BUILD_DIR=$(xcodebuild -showBuildSettings -project "$APP_PATH.xcodeproj" -json | jq -r ".[0].buildSettings.BUILD_DIR")
+
+# Prepare the demo app
+MODE="Debug"
```
Make it Release by default?

huydhn replied:
I try it out and think let's just stick with debug mode for now as it's the default one that xcodebuild uses, and it's also referred to in the test spec.  Maybe there is some benefit when it comes to benchmark in release build, but it's hard to say now

guangy10 reviewed (commented):
The workflow looks good to me.

huydhn committed:
Try to look for the model in aatp/data

huydhn committed:
Use the default debug mode from xcodebuild

huydhn committed:
Add TODO for debug mode

huydhn committed:
Remove debug message

guangy10 reviewed (approved):
Looks great to me. 

@shoumikhin can you take a look so we can unblock this PR?

huydhn committed:
Add LLaMA perf benchmark workflow for Apple iOS (#4953)

Summary:
This PR setups the framework to run LLaMA perf benchmark on Apple iOS:

* LLaMAPerfBenchmarkApp is an empty app just to host the benchmark on device.  We could extend this later to make it the generic perf benchmark loader app for iOS
* The model (llama2) is benchmarked in `testLlama2`.  I follow the same example from ET demo app where another model (mv3) is bundled together with the test suite

One caveat that I need to figure out is how to get the benchmark results out. Atm, I use this script https://github.com/ChrisMash/XCResultExtractor to extract it manually from `xcresult` file. For example,

```
$ python xcresult_extractor.py --file /Users/huydo/Library/Developer/Xcode/DerivedData/LLaMA-bchtdulhbuqeyparjpfpuchdxbva/Logs/Test/Test-LLaMAPerfBenchmark-2024.08.29_14-37-08--0700.xcresult

Test Suite 'LLaMAPerfBenchmarkTests' started at 2024-08-29 14:38:54.228.
Test Case '-[LLaMAPerfBenchmarkTests.LLaMAPerfBenchmarkTests testLlama2]' started.
2024-08-29 14:38:54.230612-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:00.559179042000000 executorch:runner.cpp:69] Creating LLaMa runner: model_path=/Users/huydo/Library/Developer/XCTestDevices/141725A5-9829-4738-95E4-5CF3A5688020/data/Containers/Bundle/Application/0EC501AC-B18F-4303-A98D-35D3DEBDDAE6/LLaMAPerfBenchmark.app/PlugIns/LLaMAPerfBenchmarkTests.xctest/llam$
2024-08-29 14:38:54.230715-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:00.559179 executorch:runner.cpp:69] Creating LLaMa runner: model_path=/Users/huydo/Library/Developer/XCTestDevices/141725A5-9829-4738-95E4-5CF3A5688020/data/Containers/Bundle/Application/0EC501AC-B18F-4303-A98D-35D3DEBDDAE6/LLaMAPerfBenchmark.app/PlugIns/LLaMAPerfBenchmarkTests.xctest/llam$
2024-08-29 14:38:54.721031-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:01.1049707332000000 executorch:runner.cpp:85] Reading metadata from model
2024-08-29 14:38:54.721134-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:01.049707 executorch:runner.cpp:85] Reading metadata from model
2024-08-29 14:38:54.721690-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:01.1050379124000000 executorch:runner.cpp:110] Metadata: get_vocab_size = 32000
2024-08-29 14:38:54.721739-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:01.050379 executorch:runner.cpp:110] Metadata: get_vocab_size = 32000
2024-08-29 14:38:54.721794-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:01.1050485249000000 executorch:runner.cpp:110] Metadata: get_bos_id = 1
2024-08-29 14:38:54.721828-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:01.050485 executorch:runner.cpp:110] Metadata: get_bos_id = 1
2024-08-29 14:38:54.721874-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:01.1050565666000000 executorch:runner.cpp:110] Metadata: use_sdpa_with_kv_cache = 1
2024-08-29 14:38:54.721923-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:01.050565 executorch:runner.cpp:110] Metadata: use_sdpa_with_kv_cache = 1
2024-08-29 14:38:54.721993-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:01.1050680207000000 executorch:runner.cpp:110] Metadata: get_n_eos = 1
2024-08-29 14:38:54.722034-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:01.050680 executorch:runner.cpp:110] Metadata: get_n_eos = 1
2024-08-29 14:38:54.722079-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:01.1050770707000000 executorch:runner.cpp:110] Metadata: append_eos_to_prompt = 0
2024-08-29 14:38:54.722120-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:01.050770 executorch:runner.cpp:110] Metadata: append_eos_to_prompt = 0
2024-08-29 14:38:54.722167-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:01.1050857957000000 executorch:runner.cpp:110] Metadata: get_max_seq_len = 128
2024-08-29 14:38:54.722578-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:01.050857 executorch:runner.cpp:110] Metadata: get_max_seq_len = 128
2024-08-29 14:38:54.722633-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:01.1051322416000000 executorch:runner.cpp:110] Metadata: enable_dynamic_shape = 1
2024-08-29 14:38:54.722677-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:01.051322 executorch:runner.cpp:110] Metadata: enable_dynamic_shape = 1
2024-08-29 14:38:54.722719-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:01.1051410124000000 executorch:runner.cpp:110] Metadata: use_kv_cache = 1
2024-08-29 14:38:54.722755-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:01.051410 executorch:runner.cpp:110] Metadata: use_kv_cache = 1
2024-08-29 14:38:54.722805-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:01.1051494791000000 executorch:runner.cpp:110] Metadata: get_n_bos = 1
2024-08-29 14:38:54.722842-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:01.051494 executorch:runner.cpp:110] Metadata: get_n_bos = 1
2024-08-29 14:38:54.722887-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:01.1051578374000000 executorch:runner.cpp:117] eos_id = 2
2024-08-29 14:38:54.722924-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:01.051578 executorch:runner.cpp:117] eos_id = 2
2024-08-29 14:38:55.363728-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:01.1692404707000000 executorch:text_prefiller.cpp:57] Prefill token result numel(): 32000
2024-08-29 14:38:55.363829-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:01.692404 executorch:text_prefiller.cpp:57] Prefill token result numel(): 32000
How do you do! I'm testing llama2 on mobile device."
Gretlasses, a yummy-looking tool, still beaming.
"Can I have one?" asked the little girl.
Gretilla nodded, and handed her sister a spinning glass.
Gretilla grabbed the glass and started to test it.
"Ah, it's so pretty!" exclaimed the little girl.
Gretilla smiled and said, "Yes, it's very pretty. Now, let's take a break and enjoy this yummy glass together!"
They both
PyTorchObserver {"prompt_tokens":16,"generated_tokens":111,"model_load_start_ms":0,"model_load_end_ms":0,"inference_start_ms":1724967534723,"inference_end_ms":1724967546771,"prompt_eval_end_ms":1724967535364,"first_token_ms":1724967535364,"aggregate_sampling_time_ms":62,"SCALING_FACTOR_UNITS_PER_SECOND":1000}
2024-08-29 14:39:06.771312-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:13.13099907195000000 executorch:stats.h:84] 	Prompt Tokens: 16    Generated Tokens: 111
2024-08-29 14:39:06.771428-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:13.099907 executorch:stats.h:84] 	Prompt Tokens: 16    Generated Tokens: 111
2024-08-29 14:39:06.771492-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:13.13100095862000000 executorch:stats.h:90] 	Model Load Time:		0.000000 (seconds)
2024-08-29 14:39:06.771542-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:13.100095 executorch:stats.h:90] 	Model Load Time:		0.000000 (seconds)
2024-08-29 14:39:06.771583-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:13.13100188904000000 executorch:stats.h:100] 	Total inference time:		12.048000 (seconds)		 Rate: 	9.213147 (tokens/second)
2024-08-29 14:39:06.771628-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:13.100188 executorch:stats.h:100] 	Total inference time:		12.048000 (seconds)		 Rate: 	9.213147 (tokens/second)
2024-08-29 14:39:06.771678-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:13.13100283195000000 executorch:stats.h:108] 		Prompt evaluation:	0.641000 (seconds)		 Rate: 	24.960998 (tokens/second)
2024-08-29 14:39:06.771733-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:13.100283 executorch:stats.h:108] 		Prompt evaluation:	0.641000 (seconds)		 Rate: 	24.960998 (tokens/second)
2024-08-29 14:39:06.771785-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:13.13100392487000000 executorch:stats.h:119] 		Generated 111 tokens:	11.407000 (seconds)		 Rate: 	9.730867 (tokens/second)
2024-08-29 14:39:06.771855-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:13.100392 executorch:stats.h:119] 		Generated 111 tokens:	11.407000 (seconds)		 Rate: 	9.730867 (tokens/second)
2024-08-29 14:39:06.771890-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:13.13100496737000000 executorch:stats.h:127] 	Time to first generated token:	0.641000 (seconds)
2024-08-29 14:39:06.771980-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:13.100496 executorch:stats.h:127] 	Time to first generated token:	0.641000 (seconds)
2024-08-29 14:39:06.772028-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:13.13100634320000000 executorch:stats.h:134] 	Sampling time over 127 tokens:	0.062000 (seconds)
2024-08-29 14:39:06.772114-0700 LLaMAPerfBenchmark[47400:2767640] I 00:00:13.100634 executorch:stats.h:134] 	Sampling time over 127 tokens:	0.062000 (seconds)
Test Case '-[LLaMAPerfBenchmarkTests.LLaMAPerfBenchmarkTests testLlama2]' passed (12.654 seconds).
Test Suite 'LLaMAPerfBenchmarkTests' passed at 2024-08-29 14:39:06.883.
	 Executed 1 test, with 0 failures (0 unexpected) in 12.654 (12.655) seconds
```

### Testing

https://github.com/pytorch/executorch/actions/runs/10639727158/job/29498590702


Reviewed By: shoumikhin, guangy10

Differential Revision: D61998563

Pulled By: huydhn

huydhn committed:
Merge branch 'main' into ios-llama-aws-device-farm

huydhn committed:
Merge branch 'ios-llama-aws-device-farm' of github.com:pytorch/executorch into ios-llama-aws-device-farm
